### PR TITLE
spatial dropout for NoisyEmbedding, LastMaxAvgEncoder, warning for bidir RnnEncoder

### PR DESCRIPTION
**Pull request by justalge**

### CODE REVIEW

File: ptls/nn/trx_encoder/trx_encoder.py [hunk 1]
```diff
@@ -60,6 +60,8 @@ def __init__(self,
                  use_batch_norm_with_lens=False,
                  clip_replace_value=None,
                  positions=None,
+                 emb_dropout=0,
```
Comment (ivkireev86) [inline ptls/nn/trx_encoder/trx_encoder.py:63]: Здесь и в других местах нужно добавить в докстринги

File: ptls/nn/trx_encoder/noisy_embedding.py [hunk 2]
```diff
@@ -25,14 +25,19 @@ def __init__(self, num_embeddings, embedding_dim, padding_idx=None, max_norm=Non
...
+        self.spatial_dropout = spatial_dropout
+        self.dropout = nn.Dropout2d(dropout) if spatial_dropout else nn.Dropout(dropout)
 
         # This is workaround for https://github.com/pytorch/pytorch/issues/44792
         # Weight are normalized after forward pass
         _ = super().forward(torch.arange(num_embeddings))
 
     def forward(self, x):
-        x = self.dropout(super().forward(x))
+        x = super().forward(x)
```
Comment (ivkireev86) [inline ptls/nn/trx_encoder/noisy_embedding.py:36]: Нужно покрыть тестами

File: ptls/nn/seq_step.py [hunk 3]
```diff
@@ -24,6 +23,15 @@ def forward(self, x: PaddedBatch):
         return h
 
 
+class LastMaxAvgEncoder(nn.Module):
```
Comment (ivkireev86) [inline ptls/nn/seq_step.py:26]: Здесь тоже докстринги с коротким примером использования

File: ptls/nn/seq_encoder/rnn_encoder.py [hunk 4]
```diff
@@ -49,14 +49,15 @@ def __init__(self,
...
+                 reducer='last_step'
                  ):
         super().__init__(is_reduce_sequence=is_reduce_sequence)
 
         self.hidden_size = hidden_size
         self.rnn_type = type
         self.bidirectional = bidir
-        if self.bidirectional:
-            raise AttributeError('bidirectional RNN is not supported yet')
+        # if self.bidirectional:
```
Comment (ivkireev86) [inline ptls/nn/seq_encoder/rnn_encoder.py:59]: Лучше не комментировать за заменить на UserWarning c предупреждением что обратное направление не совсем корректно учитывает порядок padding values.

**Comment by ivkireev86** [comment]
@justalge ,  еще просьба поменять название реквеста, чтобы во время squash&merge получился коммит с названием, отражающим суть изменений.

File: ptls/nn/seq_encoder/rnn_encoder.py [hunk 5]
```diff
@@ -66,14 +72,16 @@ def __init__(self,
                 self.hidden_size,
                 num_layers=1,
                 batch_first=True,
-                bidirectional=self.bidirectional)
+                bidirectional=self.bidirectional,
+                dropout=dropout)
```
Comment (ivkireev86) [inline ptls/nn/seq_encoder/rnn_encoder.py:76]: `dropout` не работает если `num_layers=1`, он между слоями выставляется.

**Review by ivkireev86 — APPROVED** [review]
